File: Spring/task3/task3/Card.cs
using System;
using System.Collections.Generic;

namespace Task3
{
    public enum Suit
    {
        Hearts,
        Diamonds,
        Clubs,
        Spades
    }

    public enum Value
    {
        Two = 2,
        Three,
        Four,
        Five,
        Six,
        Seven,
        Eight,
        Nine,
        Ten,
        Jack,
        Queen,
        King,
        Ace
    }

    public struct Card
    {
        internal Suit Suit;
        internal Value Value;

        public override string ToString()
        {
            return $"{Value} of {Suit}";
        }

        public uint Score()
        {
            if (Value == Value.Ace)
            {
                return 11;
            }
            if (Value <= Value.Ten)
            {
                return (uint)Value;
            }
            return 10;
        }

        public static uint GetScore(IEnumerable<Card> list)
        {
            uint result = 0;
            uint aces = 0;
            foreach (var card in list)
            {
                if (card.Value == Value.Ace)
                {
                    aces++;
                }
                else
                {
                    result += card.Score();
                }
            }

            if (result + aces > 21)
            {
                return result + aces;
            }

            // Now, total aces can give
            //     aces + M * 10
            // where M in [0..aces]
            // Hence, score is
            //     result + aces + M * 10
            // We want this value to be <= 21
            //     result + aces + M * 10 <= 21
            //     M <= (21 - result - aces) / 10
            
            uint m = (21 - result - aces) / 10;

            if (m > aces)
            {
                m = aces;
            }
            
            return result + aces + m * 10;
        }

        public static bool IsPair(List<Card> hand)
        {
            return hand.Count == 2 && hand[0].Score() == hand[1].Score();
        }
    }
}

File: Spring/Task 3 (Blackjack)/src/Card.java
public class Card implements Copyable<Card> {
    private Value value;
    private Suit suit;
    
    public Card(Value value, Suit suit) {
        this.value = value;
        this.suit = suit;
    }
    
    public Card(Card copied) {
        this.value = copied.value;
        this.suit = copied.suit;
    }
    
    @Override
    public Card copy() {
        return new Card(this);
    }
    
    @Override
    public void copyTo(Card target) {
        target.value = this.value;
        target.suit = this.suit;
    }
    
    public Value getValue() {
        return value;
    }
    
    public Suit getSuit() {
        return suit;
    }
    
    public enum Value {
        TWO(2),
        THREE(3),
        FOUR(4),
        FIVE(5),
        SIX(6),
        SEVEN(7),
        EIGHT(8),
        NINE(9),
        TEN(10),
        JACK(10),
        QUEEN(10),
        KING(10),
        ACE(11);
        
        private final int score;
        
        Value(int score) {
            this.score = score;
        }
        
        public int getScore() {
            return score;
        }
    }
    
    public enum Suit {
        HEARTS,
        DIAMONDS,
        CLUBS,
        SPADES
    }
    
    @Override
    public String toString() {
        return "{" + value + "," + suit + "}";
    }
}
